### ISSUE TITLE
Make resume_or_load support machines without a shared filesystem

### ISSUE THREAD

**Issue by MARMOTatZJU** [facebookresearch/detectron2#1732]
## Instructions To Reproduce the 🐛 Bug:

1. what changes you made (`git diff`) or what code you wrote
```
No change to code
```
2. what exact command you run: 

I performed 2 machines 4 gpus (thus 2 gpus per machine) training. Firstly, on each machine, I run the following code and change machine-rank for each machine

```bash
./tools/train_net.py --config-file configs/COCO-Detection/retinanet_R_50_FPN_3x.y
aml --num-gpu 2 --num-machines 2 --machine-rank 0 --dist-url "tcp://XXX.XXX.XXX.XXX:XXX" 
```

Then I killed the training for some reasons and later resumed the training with the following code.
```bash
./tools/train_net.py --config-file configs/COCO-Detection/retinanet_R_50_FPN_3x.y
aml --resume --num-gpu 2 --num-machines 2 --machine-rank 0 --dist-url "tcp://XXX.XXX.XXX.XXX:XXX" 
```

3. what you observed (including __full logs__):

I observed that the training loss got a sudden rise after resuming. I wonder why the loss get so high after resuming.
![image](https://user-images.githubusercontent.com/32362315/86777970-67580e80-c08c-11ea-93f7-0889fe15ce98.png)

**[UPDATE]**
full logs of all ranks (four in total):
[log.txt](https://github.com/facebookresearch/detectron2/files/4887880/log.txt)
[log.txt.rank1.txt](https://github.com/facebookresearch/detectron2/files/4887882/log.txt.rank1.txt)
[log.txt.rank2.txt](https://github.com/facebookresearch/detectron2/files/4887883/log.txt.rank2.txt)
[log.txt.rank3.txt](https://github.com/facebookresearch/detectron2/files/4887885/log.txt.rank3.txt)

4. please simplify the steps as much as possible so they do not require additional resources to
	 run, such as a private dataset.

## Expected behavior:

If there are no obvious error in "what you observed" provided above,
please tell us the expected behavior.

## Environment:

Provide your environment information using the following command:
```
wget -nc -q https://github.com/facebookresearch/detectron2/raw/master/detectron2/utils/collect_env.py && python collect_env.py
```

```
----------------------  ---------------------------------------------------------------------
sys.platform            linux
Python                  3.6.9 (default, Apr 18 2020, 01:56:04) [GCC 8.4.0]
numpy                   1.19.0
detectron2              0.1.3 @/***/detectron2
Compiler                GCC 7.5
CUDA compiler           CUDA 10.2
detectron2 arch flags   sm_60
DETECTRON2_ENV_MODULE   <not set>
PyTorch                 1.5.1 @/usr/local/lib/python3.6/dist-packages/torch
PyTorch debug build     False
GPU available           True
GPU 0,1                 Tesla P100-PCIE-16GB
CUDA_HOME               /usr/local/cuda
Pillow                  7.1.2
torchvision             0.6.1 @/usr/local/lib/python3.6/dist-packages/torchvision
torchvision arch flags  sm_35, sm_50, sm_60, sm_70, sm_75
fvcore                  0.1.1.post20200704
cv2                     4.2.0
----------------------  ---------------------------------------------------------------------
PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2019.0.5 Product Build 20190808 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.21.1 (Git Hash 7d2fd500bc78936d1d648ca713b901012f470dbc)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 10.2
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;$
rch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_37,code=compute_37
  - CuDNN 7.6.5
  - Magma 2.5.2
  - Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DNDEBUG -DUSE_FBGEMM -DUSE_QNNP$
CK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DUSE_INTERNAL_THREADPOOL_IMPL -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-fiel$
-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-func
tion -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wn
o-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -f
no-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=ON, USE_EXC
EPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=
OFF,


```

If your issue looks like an installation issue / environment issue,
please first try to solve it yourself with the instructions in
https://detectron2.readthedocs.io/tutorials/install.html#common-installation-issues

**Comment by rds-itga** [comment]
Hi, if you use Base-RetinaNet.yaml and configs/COCO-Detection/retinanet_R_50_FPN_3x.y there is no default checkpoint period defined, i think you should fill this field in your config file: SOLVER.CHECKPOINT_PERIOD.
With the default config he has no checkpoint to start from with --resume so he re-starts from the beginning.

Hope it could help you

**Comment by MARMOTatZJU** [comment]
@rds-itga Hi. Thanks a lot for your answer. Unfortunately, it doesn't work for me. _total_loss_ still starts from 1.00+ comparing to the value of 0.5 before resuming. Would you mind offer some more info about this issue in detail? Thank you in advance.
```
# latest checkpoint was at epoch 129999
./tools/train_net.py --config-file configs/COCO-Detection/retinanet_R_5[608/1815]
aml --resume --num-gpu 2 --num-machines 2 --machine-rank 0 --dist-url "tcp://XXX.XXX.XXX.XXX:XXX" SOLVER.CHECKPOINT_PERIOD 129999
```
BTW, besides of SOLVER.CHECKPOINT_PERIOD, do I also need to modifiy the text in _output/last_checkpoint_ so that it points to my desired latest checkpointed epoch?

**Comment by ppwwyyxx** [comment]
Please provide full logs as the issue template asked, for both commands in the issue.

**Comment by MARMOTatZJU** [comment]
@ppwwyyxx Full logs (4 ranks in total) have been uploaded. Please check them out.

> **[UPDATE]**
> full logs of all ranks (four in total):
> [log.txt](https://github.com/facebookresearch/detectron2/files/4887880/log.txt)
> [log.txt.rank1.txt](https://github.com/facebookresearch/detectron2/files/4887882/log.txt.rank1.txt)
> [log.txt.rank2.txt](https://github.com/facebookresearch/detectron2/files/4887883/log.txt.rank2.txt)
> [log.txt.rank3.txt](https://github.com/facebookresearch/detectron2/files/4887885/log.txt.rank3.txt)

**Comment by ppwwyyxx** [comment]
From the logs you can see that only the first machine loads the checkpoint and the other machine does not. I think this is because the workers in your setup do not share a filesystem.

This can be improved by always broadcasting the weights in [resume_or_load](https://detectron2.readthedocs.io/modules/engine.html#detectron2.engine.defaults.DefaultTrainer.resume_or_load) from the first process.

**Comment by MARMOTatZJU** [comment]
@ppwwyyxx All right, this explains the problem. Thanks a lot for your answer.

In my opinion, “broadcasting weight from the leader to the follower workers" sounds more likely to be the default behavior. Also, "not having a shared filesystem" is due to the demand of utilizing multiple independent machines which could be common to a portion of users like me, isn't it?

To this point, could you please offer any concrete advice for me to temporarily modify the code so that I can broadcast the weights from my leader worker to other follower workers while reloading the weights of a checkpoint? Thanks in advance!

**Comment by yellowjs0304** [comment]
@MARMOTatZJU 
Hi, It has been too long to talk about this, but I have the same issue as you. So, what should I do to resume the last checkpoints?
